hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-router/src/main/java/org/apache/hadoop/yarn/server/router/clientrm/RouterYarnClientUtils.java

```diff
@@ -52,4 +65,110 @@ public static GetClusterMetricsResponse merge(
+          Collection<GetApplicationsResponse> responses,
+          boolean returnPartialResult){
+    Map<ApplicationId, ApplicationReport> federationAM = new HashMap<>();
+    Map<ApplicationId, ApplicationReport> federationUAMSum = new HashMap<>();
+
+    for(GetApplicationsResponse appResponse : responses){
+      for(ApplicationReport appReport : appResponse.getApplicationList()){
+        ApplicationId appId = appReport.getApplicationId();
+        // Check if this ApplicationReport is an AM
+        if (appReport.getHost() != null) {
+          // Insert in the list of AM
+          federationAM.put(appId, appReport);
+          // Check if there are any UAM found before
+          if (federationUAMSum.containsKey(appId)) {
+            // Merge the current AM with the found UAM
+            mergeAMWithUAM(appReport, federationUAMSum.get(appId));
+            // Remove the sum of the UAMs
+            federationUAMSum.remove(appId);
+          }
+          // This ApplicationReport is an UAM
+        } else {
+          if (federationAM.containsKey(appId)) {
+            // Merge the current UAM with its own AM
+            mergeAMWithUAM(federationAM.get(appId), appReport);
+          } else if (federationUAMSum.containsKey(appId)) {
+            // Merge the current UAM with its own UAM and update the list of UAM
+            federationUAMSum.put(appId,
+                    mergeUAMWithUAM(federationUAMSum.get(appId), appReport));
+          } else {
+            // Insert in the list of UAM
+            federationUAMSum.put(appId, appReport);
+          }
+        }
+      }
+    }
+    // Check the remaining UAMs are depending or not from federation
+    for (ApplicationReport appReport : federationUAMSum.values()) {
+      if (returnPartialResult || appReport.getName() != null
+              && !(appReport.getName()
+              .startsWith(UnmanagedApplicationManager.APP_NAME)
+              || appReport.getName()
+              .startsWith(PARTIAL_REPORT))) {
+        federationAM.put(appReport.getApplicationId(), appReport);
+      }
+    }
+
+    List<ApplicationReport> appList = new ArrayList<>(federationAM.values());
+    return GetApplicationsResponse.newInstance(appList);
+  }
+
+  private static ApplicationReport mergeUAMWithUAM(ApplicationReport uam1,
+                                                   ApplicationReport uam2){
+    uam1.setName(PARTIAL_REPORT + uam1.getApplicationId());
+    mergeAMWithUAM(uam1, uam1);
+    mergeAMWithUAM(uam1, uam2);
+    return uam1;
+  }
+
+  private static void mergeAMWithUAM(ApplicationReport am,
+                                     ApplicationReport uam){
+    ApplicationResourceUsageReport resourceUsageReport =
+            am.getApplicationResourceUsageReport();
+    resourceUsageReport.setNumUsedContainers(
+            resourceUsageReport.getNumUsedContainers() +
+                    uam.getApplicationResourceUsageReport()
```

Comment:
The readability is not very good here.
Maybe extracting the report for uam would help.

Reply:
Extracted uamresourceReport to make it more readable.